### ISSUE TITLE
torch.hub.load and detect.py

### ISSUE THREAD

**Issue by protestToViolence** [ultralytics/yolov5#10364]
### Search before asking

- [X] I have searched the YOLOv5 [issues](https://github.com/ultralytics/yolov5/issues) and [discussions](https://github.com/ultralytics/yolov5/discussions) and found no similar questions.

### Question

I am using following in streamlit:
 model = torch.hub.load('yolov5', 'custom', path='best.pt',source=''local',force_reload = False)
It must be using detect.py function of the yolov5 repository. But when i actually use detect.py from the yolov5 folder using my customized training model and passing parameters i get different results.
But when i use it from torch.hub.load i dont get that results as above.
Please can somebody explain me what is happening inside torch.hub.load

**Comment by glenn-jocher** [comment]
👋 Hello, thanks for asking about the differences between [train.py](https://github.com/ultralytics/yolov5/blob/master/train.py), [detect.py](https://github.com/ultralytics/yolov5/blob/master/detect.py) and [val.py](https://github.com/ultralytics/yolov5/blob/master/val.py) in YOLOv5 🚀.

These 3 files are designed for different purposes and utilize different dataloaders with different settings. train.py dataloaders are designed for a speed-accuracy compromise, val.py is designed to obtain the best mAP on a validation dataset, and detect.py is designed for best real-world inference results. A few important aspects of each:


## train.py

- **trainloader** LoadImagesAndLabels(): designed to load train dataset images and labels. Augmentation capable and enabled. https://github.com/ultralytics/yolov5/blob/7ee5aed0b3fa6a805afb0a820c40bbcbf29960de/train.py#L210-L213 **val_loader** LoadImagesAndLabels(): designed to load val dataset images and labels. Augmentation capable but disabled. https://github.com/ultralytics/yolov5/blob/7ee5aed0b3fa6a805afb0a820c40bbcbf29960de/train.py#L220-L223
- image size: `640`
- rectangular inference: `False`
- [confidence threshold](https://github.com/ultralytics/yolov5/blob/16206692f245e713a5beb380de6dc4bed944986c/test.py#L291): `0.001`
- [iou threshold](https://github.com/ultralytics/yolov5/blob/16206692f245e713a5beb380de6dc4bed944986c/test.py#L292): `0.6`
- [multi-label](https://github.com/ultralytics/yolov5/blob/16206692f245e713a5beb380de6dc4bed944986c/test.py#L120): `True`
- [padding](https://github.com/ultralytics/yolov5/blob/16206692f245e713a5beb380de6dc4bed944986c/test.py#L83-L90): `None`

## val.py

- **dataloader** LoadImagesAndLabels(): designed to load train, val, test dataset images and labels. Augmentation capable but disabled. 
- https://github.com/ultralytics/yolov5/blob/7ee5aed0b3fa6a805afb0a820c40bbcbf29960de/val.py#L152-L153
- image size: `640`
- rectangular inference: `True`
- [confidence threshold](https://github.com/ultralytics/yolov5/blob/16206692f245e713a5beb380de6dc4bed944986c/test.py#L291): `0.001`
- [iou threshold](https://github.com/ultralytics/yolov5/blob/16206692f245e713a5beb380de6dc4bed944986c/test.py#L292): `0.6`
- [multi-label](https://github.com/ultralytics/yolov5/blob/16206692f245e713a5beb380de6dc4bed944986c/test.py#L120): `True`
- [padding](https://github.com/ultralytics/yolov5/blob/16206692f245e713a5beb380de6dc4bed944986c/test.py#L83-L90): `0.5 * maximum stride`


## detect.py

- **dataloaders** (multiple): designed for loading multiple types of media (images, videos, globs, directories, streams). https://github.com/ultralytics/yolov5/blob/7ee5aed0b3fa6a805afb0a820c40bbcbf29960de/detect.py#L120-L128
- image size: `640`
- rectangular inference: `True`
- [confidence threshold](https://github.com/ultralytics/yolov5/blob/16206692f245e713a5beb380de6dc4bed944986c/detect.py#L152): `0.25` 
- [iou threshold](https://github.com/ultralytics/yolov5/blob/16206692f245e713a5beb380de6dc4bed944986c/detect.py#L153): `0.45`
- [multi-label](https://github.com/ultralytics/yolov5/blob/16206692f245e713a5beb380de6dc4bed944986c/utils/general.py#L415): `False`
- [padding](https://github.com/ultralytics/yolov5/blob/16206692f245e713a5beb380de6dc4bed944986c/detect.py#L45-L54): `None`


## YOLOv5 PyTorch Hub Inference

YOLOv5 PyTorch Hub models are`AutoShape()` instances used for image loading, preprocessing, inference and NMS. For more info see [YOLOv5 PyTorch Hub Tutorial](https://docs.ultralytics.com/yolov5/tutorials/pytorch_hub_model_loading)
https://github.com/ultralytics/yolov5/blob/7ee5aed0b3fa6a805afb0a820c40bbcbf29960de/models/common.py#L276-L282

- image size: `640`
- rectangular inference: `True`
- [confidence threshold](https://github.com/ultralytics/yolov5/blob/abfcf9eb79877971acd238cafe6149711c5056ad/models/common.py#L226-L230): `0.25 `
- [iou threshold](https://github.com/ultralytics/yolov5/blob/abfcf9eb79877971acd238cafe6149711c5056ad/models/common.py#L226-L230): `0.45`
- [multi-label](https://github.com/ultralytics/yolov5/blob/abfcf9eb79877971acd238cafe6149711c5056ad/models/common.py#L288): `False`
- [padding](https://github.com/ultralytics/yolov5/blob/abfcf9eb79877971acd238cafe6149711c5056ad/models/common.py#L226-L230): `None`

Good luck 🍀 and let us know if you have any other questions!

**Comment by StephenZhao1** [comment]
> 👋 Hello, thanks for asking about the differences between [train.py](https://github.com/ultralytics/yolov5/blob/master/train.py), [detect.py](https://github.com/ultralytics/yolov5/blob/master/detect.py) and [val.py](https://github.com/ultralytics/yolov5/blob/master/val.py) in YOLOv5 🚀.
> 
> These 3 files are designed for different purposes and utilize different dataloaders with different settings. train.py dataloaders are designed for a speed-accuracy compromise, val.py is designed to obtain the best mAP on a validation dataset, and detect.py is designed for best real-world inference results. A few important aspects of each:
> 
> ## train.py
> * **trainloader** LoadImagesAndLabels(): designed to load train dataset images and labels. Augmentation capable and enabled. https://github.com/ultralytics/yolov5/blob/7ee5aed0b3fa6a805afb0a820c40bbcbf29960de/train.py#L210-L213
>    **val_loader** LoadImagesAndLabels(): designed to load val dataset images and labels. Augmentation capable but disabled. https://github.com/ultralytics/yolov5/blob/7ee5aed0b3fa6a805afb0a820c40bbcbf29960de/train.py#L220-L223
> * image size: `640`
> * rectangular inference: `False`
> * [confidence threshold](https://github.com/ultralytics/yolov5/blob/16206692f245e713a5beb380de6dc4bed944986c/test.py#L291): `0.001`
> * [iou threshold](https://github.com/ultralytics/yolov5/blob/16206692f245e713a5beb380de6dc4bed944986c/test.py#L292): `0.6`
> * [multi-label](https://github.com/ultralytics/yolov5/blob/16206692f245e713a5beb380de6dc4bed944986c/test.py#L120): `True`
> * [padding](https://github.com/ultralytics/yolov5/blob/16206692f245e713a5beb380de6dc4bed944986c/test.py#L83-L90): `None`
> 
> ## val.py
> * **dataloader** LoadImagesAndLabels(): designed to load train, val, test dataset images and labels. Augmentation capable but disabled.
> * https://github.com/ultralytics/yolov5/blob/7ee5aed0b3fa6a805afb0a820c40bbcbf29960de/val.py#L152-L153
> * image size: `640`
> * rectangular inference: `True`
> * [confidence threshold](https://github.com/ultralytics/yolov5/blob/16206692f245e713a5beb380de6dc4bed944986c/test.py#L291): `0.001`
> * [iou threshold](https://github.com/ultralytics/yolov5/blob/16206692f245e713a5beb380de6dc4bed944986c/test.py#L292): `0.6`
> * [multi-label](https://github.com/ultralytics/yolov5/blob/16206692f245e713a5beb380de6dc4bed944986c/test.py#L120): `True`
> * [padding](https://github.com/ultralytics/yolov5/blob/16206692f245e713a5beb380de6dc4bed944986c/test.py#L83-L90): `0.5 * maximum stride`
> 
> ## detect.py
> * **dataloaders** (multiple): designed for loading multiple types of media (images, videos, globs, directories, streams). https://github.com/ultralytics/yolov5/blob/7ee5aed0b3fa6a805afb0a820c40bbcbf29960de/detect.py#L120-L128
> * image size: `640`
> * rectangular inference: `True`
> * [confidence threshold](https://github.com/ultralytics/yolov5/blob/16206692f245e713a5beb380de6dc4bed944986c/detect.py#L152): `0.25`
> * [iou threshold](https://github.com/ultralytics/yolov5/blob/16206692f245e713a5beb380de6dc4bed944986c/detect.py#L153): `0.45`
> * [multi-label](https://github.com/ultralytics/yolov5/blob/16206692f245e713a5beb380de6dc4bed944986c/utils/general.py#L415): `False`
> * [padding](https://github.com/ultralytics/yolov5/blob/16206692f245e713a5beb380de6dc4bed944986c/detect.py#L45-L54): `None`
> 
> ## YOLOv5 PyTorch Hub Inference
> YOLOv5 PyTorch Hub models are`AutoShape()` instances used for image loading, preprocessing, inference and NMS. For more info see [YOLOv5 PyTorch Hub Tutorial](https://github.com/ultralytics/yolov5/issues/36)
> 
> https://github.com/ultralytics/yolov5/blob/7ee5aed0b3fa6a805afb0a820c40bbcbf29960de/models/common.py#L276-L282
> 
> * image size: `640`
> * rectangular inference: `True`
> * [confidence threshold](https://github.com/ultralytics/yolov5/blob/abfcf9eb79877971acd238cafe6149711c5056ad/models/common.py#L226-L230): `0.25 `
> * [iou threshold](https://github.com/ultralytics/yolov5/blob/abfcf9eb79877971acd238cafe6149711c5056ad/models/common.py#L226-L230): `0.45`
> * [multi-label](https://github.com/ultralytics/yolov5/blob/abfcf9eb79877971acd238cafe6149711c5056ad/models/common.py#L288): `False`
> * [padding](https://github.com/ultralytics/yolov5/blob/abfcf9eb79877971acd238cafe6149711c5056ad/models/common.py#L226-L230): `None`
> 
> Good luck 🍀 and let us know if you have any other questions!

python detect.py --weights **.pth --source ***.jpg results is totally different from torch.hub.load with same weights file, how does it happen?

**Comment by StephenZhao1** [comment]
> ### Search before asking
> * [x]  I have searched the YOLOv5 [issues](https://github.com/ultralytics/yolov5/issues) and [discussions](https://github.com/ultralytics/yolov5/discussions) and found no similar questions.
> 
> ### Question
> I am using following in streamlit: model = torch.hub.load('yolov5', 'custom', path='best.pt',source=''local',force_reload = False) It must be using detect.py function of the yolov5 repository. But when i actually use detect.py from the yolov5 folder using my customized training model and passing parameters i get different results. But when i use it from torch.hub.load i dont get that results as above. Please can somebody explain me what is happening inside torch.hub.load
> 
> ### Additional
> _No response_

When you use opencv read image, the color channel is BGR,  the input of torch model is RGB,
I change code like this, image = cv2.imread(img_path)[:, :, ::-1]

**Comment by glenn-jocher** [comment]
@StephenZhao1 got it! It seems like there was a mismatch in the color channels. Your modification should resolve the issue by converting the BGR format to RGB. Let me know if you need further assistance!
